### DISCUSSION TITLE
4051 with STM32F411CE

### DISCUSSION THREAD

**Discussion by gitrimius** [shanteacontrols/OpenDeck#40] [Q&A]
Hi,
I was wondering whether 4051 multiplexers are supported with the STM32F411CE board.
I have tried it out and at first sight it does not seem to work. Only pot 0 (of 0-7) reacts. I have used following analog setup in yml:

_analog:
    type: "4051"
    multiplexers: 1
    pins:
      s0:
        port: "A"
        index: 5
      s1:
        port: "A"
        index: 6
      s2:
        port: "A"
        index: 7
      z0:
        port: "A"
        index: 1_

If not supported, that is ok, but maybe I am doing something wrong?
grts!

**Comment by paradajz** [comment]
It's clearly stated in the wiki that 4051 is supported. https://github.com/paradajz/OpenDeck/wiki/Creating-custom-board-variant
4051 has 8 inputs so I don't understand what's your issue here.

**Comment by gitrimius** [comment]
yeah, I know, I had read that, but I think I was confused since I read your blog post on the mix board for Arduino. And since it was not working... 
The thing is that I only get normal readings from the first input on my 4051, the rest of the inputs don't react and interfere with input 1.
I looked into it a bit more. I notice that when I edit the yml and enter 2 for the amount of multiplexers (and an extra 'z') and flash the firmware (through the web interface) I don't get more analog inputs in the UI. I would suspect to get at least 16? I must say I did not connect 2 mux's, but that doesn't matter, no? 
Anyhow, thanks a lot for your time...

**Reply by paradajz** [reply]
Every time you change the YAML file you should run make clean

**Reply by gitrimius** [reply]
thanks for the tip!
I just tried that, but it does not make the 4051 work after rebuilding and flashing the fw through the web-interface.
It looks like all I change in the YAML file does not reflect in the web-interface..

**Reply by paradajz** [reply]
Try running factory reset from WebUI.

**Reply by gitrimius** [reply]
sorry, that didn't change the behaviour..
For my understanding: am I right to suppose that changing the YAML file should be reflected in the webUI?
thanks!

**Reply by gitrimius** [reply]
I noticed something strange just now. The webUI (Vmware Ubuntu 20.10) Chrome gave me readouts for Analog Channel 0 without the board being connected... It showed the same behaviour as with the board connected. But if I switch between grid and table view it 'looses connection'...
I must say that the virtual machine has been suspended and unsuspended before..

**Reply by paradajz** [reply]
Please type a list of *exact* steps that you are doing. You should

1) Clean up the entire repo (best by running `git clean -dxf`)
2) Run make TARGET=your-target-name
3) Flash the merged binary located in src/build/merged/your-target-name.hex

**Reply by gitrimius** [reply]
Hi,
okay, all of a sudden it works..
I tried flashing the .hex file with STM32CubeProgrammer and the 4051 works.
I think that if I flash with webUI (Ubuntu w. Chrome on VMware & Catalina w. Chrome)something isn't handled correctly. So I tried again. It is not working...
Could it be that the WebUI is not flashing at all? or did you change something in the last hour on the webUI?
I noticed following change in the UI: 

> DIN MIDI state
> Not supported on this device.
> Enable or disable DIN MIDI input and output.

**Reply by paradajz** [reply]
WebUI update works correctly. The question is - which .sysex did you gave the WebUI to update? If you want to create new .sysex file for your own board, you should run make TARGET=your-target-name sysexfw

**Reply by gitrimius** [reply]
ok, that is the part I was missing. I flashed the simple .hex file... Which one do I have to flash?

blackpill411_4051.bin
blackpill411_4051.hex
blackpill411_4051.sysex
blackpill411_4051_sysex.bin
blackpill411_4051_sysex.hex

**Reply by paradajz** [reply]
Flash: blackpill411_4051.hex
Update via WebUI: blackpill411_4051.sysex

I will cleanup this naming scheme later today.

**Reply by gitrimius** [reply]
great! Thank you very much for the quick help. I am up and running. Great project!

**Reply by paradajz** [reply]
Cleaned up and pushed. Merged build directory now follows the same structure as all the other build directories and now there is only a single .hex and single .bin file. You always flash .hex - .bin is only for my own verification purposes. Running sysexfw target via make will create additional .sysex file.

**Reply by paradajz** [reply]
Also I've removed all binaries from repository - next release will have them attached to release.

**Reply by paradajz** [reply]
Final point, .sysex creation is now mentioned here: https://github.com/paradajz/OpenDeck/wiki/Creating-custom-board-variant#building-custom-configuration

**Reply by gitrimius** [reply]
great, thanks for the help!
